### PR TITLE
network: update IpList as a list of CidrRanges

### LINKED ISSUE #1045: replace IpList with CidrRange
CidrRange seems to support both IPv4 and IPv6. Is the intention to replace IpList with CidrRange?

### PR DESCRIPTION
**Pull request by hennna**
This PR updates IpList as a list of CidrRanges.

IpList was moved from network/utility.h to network/cidr_range.h. This was done to avoid cyclic dependencies between network/utility.cc and network/cidr_range.cc. 

IpList has been moved to the Network::Address namespace from the Network:: namespace. This seems to make sense given that IpList is a list of cidr ranges. 

Additionally tests were moved from network/utility_test to network/cidr_range_test.

Fixes #1045 .

### CODE REVIEW

File: source/common/network/cidr_range.cc [hunk 1]
```diff
@@ -79,13 +79,20 @@ IpVersion CidrRange::version() const {
...
+}
+
+bool CidrRange::isInRange(const Instance& address) const {
+  if (length_ < 0 || address.type() != Type::Ip || address.ip()->version() != version()) {
     return false;
   }
   // Make an CidrRange from the address, of the same length as this. If the two ranges have
   // are the same, then the address is in this range.
-  CidrRange other = create(address, length_);
+  CidrRange other = create(address.ip()->addressAsString(), length_);
```
Comment (htuch) [inline source/common/network/cidr_range.cc]: This seems a bit inefficient, as the code is converting to a string from an address and then immediately back in `Address`. Can you modify `CidrRange` to take an address?

File: source/common/network/cidr_range.h [hunk 2]
```diff
@@ -117,6 +120,23 @@ class CidrRange {
...
 };
 
+/**
+ * Class for keeping a list of CidrRanges, and then determining whether an
+ * IP address is in the CidrRange list.
+ */
+class IpList {
+public:
+  IpList(const std::vector<std::string>& subnets);
+  IpList(const Json::Object& config, const std::string& member_name);
```
Comment (htuch) [inline source/common/network/cidr_range.h]: Should this just be a static factory? It seems that logically the only constructor we should have is the one that takes the subnets.

File: source/common/network/cidr_range.cc (hunk 1 above)
Comment (hennna) [inline source/common/network/cidr_range.cc]: I think the underlying problem is that CidrRange consists of an InstanceConstSharedPtr. Eventually the conversion from the Instance object to shared pointer must be made. It can happen in create, truncateIpAddressAndLength, or in the CidrRange constructor.

File: source/common/network/cidr_range.h (hunk 2 above)
Comment (hennna) [inline source/common/network/cidr_range.h]: Constructor is used here: https://github.com/lyft/envoy/blob/master/source/common/filter/auth/client_ssl.cc#L29. I think constructing from the json object does make sense. But changing to a static factory isn't much work.

File: source/common/network/cidr_range.cc (hunk 1 above)
Comment (htuch) [inline source/common/network/cidr_range.cc]: Sure, whichever makes sense, but it seems we should be able to do better than converting to/from a string for an operation that is essentially about copying an address and masking some bits. Of course, given the `friendly_address_` representation in string form that is maintained today, this might not be that much more efficient, but at least from a logical point of view we aren't mucking around with strings.

File: source/common/network/cidr_range.h (hunk 2 above)
Comment (htuch) [inline source/common/network/cidr_range.h]: I don't feel that strongly (I see that this was there before), but I tend to feel that the principle of minimalism would dictate this be a static factory, since you can express the JSON variation in terms of the existing constructor, it doesn't depend interact with other method or state in the object.

Comment (hennna) [inline source/common/network/cidr_range.h]: Yes. The diff here is just the file change.

File: source/common/network/cidr_range.cc (hunk 1 above)
Comment (hennna) [inline source/common/network/cidr_range.cc]: The string / socket to shared pointer conversion can happen further along in the code but the way I'm doing it here is simplest in terms of number of line changes.

Comment (htuch) [inline source/common/network/cidr_range.cc]: If you're not going to make the change in this PR, can you add a `TODO(hennna): ...` here? Personally I think it's crazy that we are munging back and forth between strings to do what are simple bitmask operations at the end of the day, but I don't feel so strongly I'll block the merge on this if a TODO is added.

Comment (hennna) [inline source/common/network/cidr_range.cc]: Let me double check with @jamessynge

File: test/common/filter/auth/client_ssl_test.cc [hunk 3]
```diff
@@ -149,8 +149,10 @@ TEST_F(ClientSslAuthFilterTest, Ssl) {
   // Create a new filter for an SSL connection, with no backing auth data yet.
   createAuthFilter();
   ON_CALL(filter_callbacks_.connection_, ssl()).WillByDefault(Return(&ssl_));
-  Network::Address::Ipv4Instance remote_address("192.168.1.1");
-  EXPECT_CALL(filter_callbacks_.connection_, remoteAddress()).WillOnce(ReturnRef(remote_address));
+  Network::Address::InstanceConstSharedPtr remote_address =
+      Network::Utility::parseInternetAddress("192.168.1.1");
+  EXPECT_CALL(filter_callbacks_.connection_, remoteAddressConstSharedPtr())
+      .WillOnce(Return(remote_address));
```
Comment (hennna) [inline test/common/filter/auth/client_ssl_test.cc]: I remember in a previous PR, I ran into some problems with testing::Return(smart pointer). Here, it seems fine I'm assuming because it's constant. @htuch can you take a second look?

File: include/envoy/network/connection.h [hunk 4]
```diff
@@ -120,14 +120,15 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
...
-
+  virtual Address::InstanceConstSharedPtr remoteAddressConstSharedPtr() PURE;
   /**
    * @return the local address of the connection. For client connections, this is the origin
    * address. For server connections, this is the local destination address. For server connections
    * it can be different from the proxy address if the downstream connection has been redirected or
    * the proxy is operating in transparent mode.
    */
   virtual const Address::Instance& localAddress() PURE;
+  virtual Address::InstanceConstSharedPtr localAddressConstSharedPtr() PURE;
```
Comment (htuch) [inline include/envoy/network/connection.h]: So the basic motivation here is to form a `CidrRange` without having to copy the underlying Address instance, as response to my review comment that we shouldn't have to convert an Address instance to string and then reparse to form an `Address:InstanceConstSharedPtr`. The reason we need to form the `CidrRange` is for some inclusion testing. Does that sound right?

It seems reasonable to me given the current way in which addresses are handled. From a pure performance PoV, I wonder if the inclusion testing could not depend on a const shared pointer in the first place? I.e. `isInRange()` shouldn't need to do this construction of `CidrRange` to begin with. That way you save on the reference count increment/decrement.

@jamessynge @mattklein123

Comment (mattklein123) [inline include/envoy/network/connection.h]: Sorry I'm a little lost. Yeah the addition of these new functions is pretty strange IMO. Why is this needed? I would need to look through all the code again, but the IPList/CidrRange construction is not performance critical. Only isContains() etc. is perf critical. Why are pointers needed at all? Can't we just pass the address reference and do the check?

File: include/envoy/network/connection.h [hunk 5]
```diff
@@ -120,6 +120,7 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
    * @return The address of the remote client.
    */
   virtual const Address::Instance& remoteAddress() PURE;
+  virtual Address::InstanceConstSharedPtr remoteAddressConstSharedPtr() PURE;
```
Comment (mattklein123) [inline include/envoy/network/connection.h]: Per hidden comment, very confused why the new functions have been added. Just putting this here so it shows up in PR.

File: include/envoy/network/connection.h (hunk 4 above)
Comment (htuch) [inline include/envoy/network/connection.h]: The issue is https://github.com/lyft/envoy/blob/master/source/common/network/cidr_range.cc#L88. If you go back to the original https://github.com/lyft/envoy/pull/1090/commits/64e5287b8504224edaf247547eba45648af2e657, this is what @hennna did. But, the implementation was via string conversion and CIDR range construction.

It strikes me (as a bit twiddler), that we should be able to perform this operation way faster than string parsing and object construction. It should just be two right shift operations and equality. This speaks to the underlying `Address::Instance` representation at EOD.

Comment (mattklein123) [inline include/envoy/network/connection.h]: I see. OK. We have to inline construct a CidrRange(). Yeah I'm not crazy about this change. Can we restructure the CidrRange code so it's possible to construct one using a reference/pointer to an Address and just be responsible for lifetime? That seems much cleaner and easier and can be private construction also since it's internal.

Comment (mattklein123) [inline include/envoy/network/connection.h]: Or just drop the duplicate construction of CidrRange and do the bit math directly as @htuch says.

Comment (htuch) [inline include/envoy/network/connection.h]: Going off-topic slightly, have we considered implementing all addresses as simply `__uint128_t`? That would greatly simplify `CidrRange` operations. @jamessynge

File: include/envoy/network/connection.h (hunk 5 above)
Comment (hennna) [inline include/envoy/network/connection.h]: IpList::contains(const Address::Instance& address) calls CidrRange::isInRange(InstanceConstSharedPtr address). 
This is trying to solve the problem of the Instance->string->InstanceConstSharedPtr conversion when calling isInRange. The way this is done is to change IpList::contains to take an InstanceConstSharedPtr instead.

Comment (mattklein123) [inline include/envoy/network/connection.h]: Right, I understand now. (See other comment thread). I think we should figure out a better solution here per the other comment thread.

File: include/envoy/network/connection.h (hunk 4 above)
Comment (mattklein123) [inline include/envoy/network/connection.h]: Yeah that seems like a different discussion. I think for this it should be possible to either skip the 2nd CidrRange construction, or to do one but pass a pointer/reference to an address as a temporary. Either seem much better than the interface changes here.

**Comment by hennna** [comment]
@htuch for review.

File: source/common/network/cidr_range.cc [hunk 6]
```diff
@@ -78,24 +78,53 @@ IpVersion CidrRange::version() const {
...
+    for (int i = 0; i < 16; i++) {
+      if (length == 0) {
+        return true;
+      } else if (length < 8) {
+        // Compare relevant bits.
+        if (address.ip()->ipv6()->address()[i] >> (8 - length) ==
+            ipv6()->address()[i] >> (8 - length)) {
+          return true;
+        } else {
+          break;
```
Comment (htuch) [inline source/common/network/cidr_range.cc]: It's probably clearer to replace `break` with `return false` here (and in this case, just replace the `if` statement with a single `return`).

File: test/common/network/cidr_range_test.cc [hunk 7]
```diff
@@ -215,10 +207,10 @@ TEST(Ipv6CidrRange, StringCtor) {
...
   EXPECT_EQ(rng.length(), 118);
   EXPECT_EQ(rng.version(), IpVersion::v6);
-  EXPECT_FALSE(rng.isInRange(Utility::parseInternetAddress("::fbff")));
-  EXPECT_TRUE(rng.isInRange(Utility::parseInternetAddress("::fc00")));
-  EXPECT_TRUE(rng.isInRange(Utility::parseInternetAddress("::ffff")));
-  EXPECT_FALSE(rng.isInRange(Utility::parseInternetAddress("::1:00")));
+  EXPECT_FALSE(rng.isInRange(Ipv6Instance("::fbff")));
+  EXPECT_TRUE(rng.isInRange(Ipv6Instance("::fc00")));
+  EXPECT_TRUE(rng.isInRange(Ipv6Instance("::ffff")));
+  EXPECT_FALSE(rng.isInRange(Ipv6Instance("::1:00")));
```
Comment (htuch) [inline test/common/network/cidr_range_test.cc]: Do these provide 100% coverage of `isinRange`? There are quite a few different branches there now..
